src/librustc_mir/build/matches/mod.rs

```diff
@@ -512,7 +512,10 @@ impl<'a,'tcx> Builder<'a,'tcx> {
         let otherwise: Vec<_> =
             target_blocks.into_iter()
                          .zip(target_candidates)
-                         .flat_map(|(target_block, target_candidates)| {
+                         .flat_map(|(target_block, mut target_candidates)| {
+                             // We need to preserve the fact that the candidates
+                             // are in the reversed order compared to the source.
+                             target_candidates.reverse();
                              self.match_candidates(span,
```

Comment:
I tried reversing just the creation of the `target_candidates` Vec but that turned out to be a lot more annoying. So, instead I just changed all the code to deal with the candidates in the same order they're presented in the source rather than reversed.